### PR TITLE
Fixed application new versions appearance in deployment job

### PR DESCRIPTION
**Pull request by rkovalenko**

### CODE REVIEW

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
+package com.epam.grandhackathon.deployment.sphere.plugin;
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: please move to dedicated package

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        return environments;
+    }
+
+	@Extension
+	public static final class DescriptorImpl extends ParameterDescriptor {
+
+		@Override
+		public String getDisplayName() {
+			return "Deploy meta data parameter";
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: extract to constant

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    	Collection<EnvironmentMetaData> envs = environmentDao.findAll();
+        List<String> environments = Lists.newArrayList();
+        for (EnvironmentMetaData env : envs) {
+            environments.add(env.getTitle());
+        }
+        return environments;
+    }
+
+	@Extension
+	public static final class DescriptorImpl extends ParameterDescriptor {
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: implementation usually is something concrete, please give more appropriate name

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		DeployMetaDataParameterValue value = req.bindJSON(DeployMetaDataParameterValue.class, jo);
+		return value;
+	}
+
+	@Override
+	public ParameterValue createValue(StaplerRequest req) {
+		return new DeployMetaDataParameterValue(getName(), this.environmentKey, this.buildVersion, this.applicationName, "Deploy meta data parameter");
+	}
+	
+	public List<String> getBuildVersions() {
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: Do we really need those methods public? if no, let's make them private. By default everything is private

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		Jenkins.getInstance().getInjector().injectMembers(this);
+		Collection<BuildMetaData> builds = buildMetaDataDao.findByAppName(applicationName);
+        List<String> versions = Lists.newArrayList();
+        for (BuildMetaData build : builds) {
+            versions.add(build.getBuildVersion());
+        }
+        return versions;
+    }
+
+    public List<String> getEnvironmentKeys() {
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: The same as above

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	@Override
+	public ParameterValue createValue(StaplerRequest req) {
+		return new DeployMetaDataParameterValue(getName(), this.environmentKey, this.buildVersion, this.applicationName, "Deploy meta data parameter");
+	}
+	
+	public List<String> getBuildVersions() {
+		Jenkins.getInstance().getInjector().injectMembers(this);
+		Collection<BuildMetaData> builds = buildMetaDataDao.findByAppName(applicationName);
+        List<String> versions = Lists.newArrayList();
+        for (BuildMetaData build : builds) {
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: use Guava transformation here

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        }
+        return versions;
+    }
+
+    public List<String> getEnvironmentKeys() {
+    	Jenkins.getInstance().getInjector().injectMembers(this);
+    	Collection<EnvironmentMetaData> envs = environmentDao.findAll();
+        List<String> environments = Lists.newArrayList();
+        for (EnvironmentMetaData env : envs) {
+            environments.add(env.getTitle());
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: use Guava transformation here

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import lombok.ToString;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.export.Exported;
+
+import com.epam.grandhackathon.deployment.sphere.plugin.metadata.Constants;
+
+@EqualsAndHashCode(callSuper = false)
+@ToString(callSuper = false)
+public class DeployMetaDataParameterValue extends ParameterValue {
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java]: move to dedicated package along with class above

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    env.put(Constants.ENV_NAME, environmentKey);
+    env.put(Constants.BUILD_VERSION, buildVersion);
+    env.put(Constants.BUILD_APP_NAME, applicationName);
+  }
+
+  @Override
+  public VariableResolver<String> createVariableResolver(AbstractBuild<?, ?> build) {
+    return new VariableResolver<String>() {
+      public String resolve(String name) {
+        return Constants.ENV_NAME.equals(name) ?
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java]: a bit complex code, please make it simpler by extracting into methods with proper names

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployVersionMetaDataPublisher.java [hunk 10]
```diff
@@ -68,64 +71,20 @@ public boolean perform(final AbstractBuild<?, ?> build, final Launcher launcher,
...
-    public boolean prebuild(final AbstractBuild<?, ?> build, final BuildListener listener) {
-        repopulateValues(build);
         return true;
     }
 
-    private void repopulateValues(final AbstractBuild<?, ?> build) {
+	private void repopulateEnvValues(final AbstractBuild<?, ?> build) {
         getDescriptor().load();
         final String appName = getDeployedAppName();
         checkState(!Strings.isNullOrEmpty(appName), String.format("Invalid application name", appName));
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployVersionMetaDataPublisher.java]: missed %s for format

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 1 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: ok.

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 2 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: ok

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 3 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: DeployMetaDataParameterDescriptorImpl - ok ?

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+			}
+		}	
+	}
+
+	private boolean isAbstractProject(Item item){
+		return item instanceof AbstractProject;
+	}
+	
+	private void addPublisherProperty(AbstractProject prj, ParametersDefinitionProperty paramDefProp, String applicationName){
+		DeployMetaDataParameterDefinition deployMetaData = new DeployMetaDataParameterDefinition(Constants.DEPLOY_META_DATA, Constants.DEPLOY_META_DATA, "", "", applicationName);
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java]: why 2 args empty? may be it's better to introduce additional constructor for such case?

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	}
+	
+	private void configureJobWithDeployVersionPublisher(Item item){
+		if (isAbstractProject(item)){
+			AbstractProject prj = (AbstractProject)item;
+			DescribableList<Publisher,Descriptor<Publisher>> publishers = prj.getPublishersList();
+			DeployVersionMetaDataPublisher dataPublisher = publishers.get(DeployVersionMetaDataPublisher.class);
+			if (dataPublisher != null){
+				ParametersDefinitionProperty paramDefProp = (ParametersDefinitionProperty)prj.getProperty(ParametersDefinitionProperty.class);
+				if (null != paramDefProp)
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java]: use curly braces for conditions, it makes read easier

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	public void onCreated(Item item) {
+		configureJobWithDeployVersionPublisher(item);	
+	}
+
+	@Override
+	public void onUpdated(Item item) {
+		configureJobWithDeployVersionPublisher(item);
+	}
+	
+	private void configureJobWithDeployVersionPublisher(Item item){
```
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java:36]: a lot of nested conditions which increases code complexity, extract into dedicated methods with proper names

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.epam.grandhackathon.deployment.sphere.plugin.listener.DeployVersionMetaDataListenerTest;
+import com.epam.grandhackathon.deployment.sphere.plugin.metadata.Constants;
+
+public class DeployMetaDataParameterValueTest {
+	private static final Logger LOGGER = LoggerFactory.getLogger(DeployVersionMetaDataListenerTest.class);
```
Comment (webdizz) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java]: u can use Lombok here as well

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 4 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: getApplicationName(),  getBuildVersions() getEnvironmentKeys() used by jelly for bind values to form. 

For example jelly script 

```
        <select name="environmentKey">
            <j:forEach var="environmentKey" items="${it.environmentKeys}">
                <option value="${environmentKey}">${environmentKey}</option>
            </j:forEach>
         </select>
```

call getEnvironmentKeys() for bind data to ui form.

Really it's jelly  requirements.

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    
+	@Rule
+    public JenkinsRule j = new JenkinsRule();
+	private String environmentKey = "qa";
+	private String buildVersion = "0.1";
+	private String applicationName = "app1";
+	
+	@Test
+	public void testBuildEnvVars() throws Exception {
```
Comment (webdizz) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java]: usually test starts with "should" e.g. shouldPopulateEnvironmentVariables for example

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListenerTest.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	@Test
+	public void testOnCreatedItem() throws Exception {
+		LOGGER.info("Create project with Deployment Sphere plugin");
+		checkProjectConfigurationWithPlugin();
+		LOGGER.info("Done.");
+		
+	}
+
+	@Test
+	public void testOnUpdatedItem() throws Exception{
```
Comment (webdizz) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListenerTest.java]: Cannot see what is being asserted

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 2 above)
Comment (webdizz) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: right, later we're going to introduce localisation

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java (hunk 8 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java]: ok

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployVersionMetaDataPublisher.java (hunk 10 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployVersionMetaDataPublisher.java]: ok

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java (hunk 9 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValue.java]: ok

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java (hunk 12 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java]: ok

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java (hunk 11 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java]: fixed

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java (hunk 13 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListener.java:36]: fixed

File: plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java (hunk 7 above)
Comment (rkovalenko) [inline plugin/src/main/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterDefinition.java]: Right? 

```
Function<EnvironmentMetaData, String> function = new Function<EnvironmentMetaData, String>() {
            @Override
            public String apply(EnvironmentMetaData data) {
                return data.getTitle();
            }
        };

Collection<String> environments = Collections2.transform(envs, function);
```

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java (hunk 14 above)
Comment (rkovalenko) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java]: Loombook @Log ?

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListenerTest.java (hunk 16 above)
Comment (rkovalenko) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/listener/DeployVersionMetaDataListenerTest.java]: fixed

File: plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java (hunk 15 above)
Comment (rkovalenko) [inline plugin/src/test/java/com/epam/grandhackathon/deployment/sphere/plugin/DeployMetaDataParameterValueTest.java]: ok

**Pull request closed without merging**